Add `Range` method

gozeloglu opened the pull request:
* This fixes #71.
* The code has not been tested yet. I need some help from you, @swithek. How can I test this code?

swithek commented:
@gozeloglu three PRs in a row, you're on fire!

As for testing, I think something like this should work:
```go
c := prepCache("1", "2", "3")
var results []string
c.Range(func(item *Item) {
    results = append(results, item.Key())
})
assert.Equal(t, []string{"1", "2", "3"}, results) // order is important
```

swithek commented on cache.go:
```diff
@@ -552,6 +552,20 @@ func (c *Cache[K, V]) OnEviction(fn func(context.Context, EvictionReason, *Item[
+
+	for k := range c.items.values {
+		item := c.get(k, false).Value.(*Item[K, V])
+		if !fn(item) {
```
It should be possible to execute other cache methods without causing any deadlocks as mentioned in this comment: https://github.com/jellydator/ttlcache/issues/71#issuecomment-1086167465

`sync.Map`'s `Range()` has this documented quite well:
`Range does not block other methods on the receiver; even f itself may call any method on m. `

gozeloglu replied:
I didn't get the main logic here. I added `mu.Lock()`-`mu.Unlock()` in `Range()`. How can I avoid deadlocks?

swithek replied:
The problem is that the user might want to use this method as follows:
```go
cache.Range(func(item *Item){
    cache.Delete(item.Key())
    cache.Set(item.Key() + "newkey", "123")
})
```

This is just an example but it illustrates my point: calling a cache method inside the range func will cause a deadlock.

The simplest solution for now would be to lock and unlock the mutex before calling the range func:
```go
func (c *Cache) Range(fn func(*Item)) {
    c.items.mu.RLock()
    for next := c.items.lru.Next(); next != nil; next = c.items.lru.Next() {
        item := next.Value.(*Item)
        c.items.mu.RUnlock()
        
        if !fn(item) {
            return
        }
        
        if c.items.lru.Next() != nil {
            c.items.mu.RLock()
        }
    }
}
```

gozeloglu replied:
I did some changes. Please let me know if i missed something.

gozeloglu commented:
Hi @swithek. Can you review the last changes?

swithek commented:
@gozeloglu sorry for the delay. I'll review your changes sometime in the next few days.

swithek commented:
> I added a simple test but I only checked the "5" because of map internal.
My proposal in the comment above uses the LRU list instead of the map, so you should be able to get predictable results and check all items in the result slice.

gozeloglu commented:
> > I added a simple test but I only checked the "5" because of map internal.
> 
> My proposal in the comment above uses the LRU list instead of the map, so you should be able to get predictable results and check all items in the result slice. @gozeloglu

I updated the test. The keys are added to the front of the linked list. So, I checked in reverse order.

swithek commented on cache.go:
```diff
@@ -552,6 +552,19 @@ func (c *Cache[K, V]) OnEviction(fn func(context.Context, EvictionReason, *Item[
+// until it returns false.
+func (c *Cache[K, V]) Range(fn func(item *Item[K, V]) bool) {
+	for item := c.items.lru.Front(); item != c.items.lru.Back().Next(); item = item.Next() {
+		c.items.mu.RLock()
```
This still has a pretty big problem: `c.items.lru` is left unprotected by the mutex. I think we need to lock it before we enter the loop and relock it before the next iteration starts.

gozeloglu replied:
You are right. I missed that point. I fixed it in b5ee285.
